File: app/src/main/java/com/quiz/translatoraalllanguage/HistoryActivity.java
package com.quiz.translatoraalllanguage;

import android.app.AlertDialog;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.speech.tts.TextToSpeech;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.Locale;

import butterknife.BindView;
import butterknife.ButterKnife;

public class HistoryActivity extends AppCompatActivity {
    CustomAdapter adapter;
    @BindView(R.id.recycler)
    RecyclerView recycler;
    String s;
    TextToSpeech t1;
    @BindView(R.id.toolbar)
    Toolbar toolbar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_history);
        ButterKnife.bind(this);

        LinearLayoutManager manager = new LinearLayoutManager(this);
        manager.setReverseLayout(true);
        manager.setStackFromEnd(true);
        recycler.setLayoutManager(manager);
        s = getIntent().getStringExtra("name");
        setSupportActionBar(toolbar);
        TextView title=toolbar.findViewById(R.id.title);
        title.setText(s.equals("Favourites.db") ? "Bookmark" : "History");
        final DataBaseHelper dataBaseHelper = new DataBaseHelper(getApplicationContext(), s);
        ArrayList<Word> arrayList = dataBaseHelper.getAllWords();
        Log.i("123321", "" + arrayList);

        adapter = new CustomAdapter(getApplicationContext(), arrayList);

        recycler.setAdapter(adapter);
        TextView clear=toolbar.findViewById(R.id.clear);
        clear.setOnClickListener(v -> {
            AlertDialog.Builder builder = new AlertDialog.Builder(this);
            builder.setTitle("Are you Sure?");
            builder.setMessage("Delete it permanently");
            builder.setPositiveButton("ok",(dialog, which) -> {
                dataBaseHelper.deleteAllWords();
                ArrayList<Word> arrayList0 = dataBaseHelper.getAllWords();


                adapter = new CustomAdapter(getApplicationContext(), arrayList0);

                recycler.setAdapter(adapter);
            } );
            builder.setNegativeButton("cancel",null );
            builder.show();
        });
    }


    public class CustomAdapter extends RecyclerView.Adapter<CustomAdapter.MyViewHolder> {

        private ArrayList<Word> originalItems;
        private ArrayList<Word> filteredItems;
        private LayoutInflater inflater;

        public CustomAdapter(@NonNull Context context,
                             @NonNull ArrayList<Word> objects) {

            this.originalItems = new ArrayList<>();
            this.originalItems.addAll(objects);
            filteredItems = new ArrayList<Word>();
            filteredItems.addAll(this.originalItems);
            inflater = LayoutInflater.from(context);
        }

        public int getCount() {
            return filteredItems.size();
        }

        public Word getItem(int position) {
            return filteredItems.get(position);
        }

        @NonNull
        @Override
        public MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View view = inflater.inflate(R.layout.list_item, parent, false);
            MyViewHolder holder = new MyViewHolder(view);
            return holder;
        }

        @Override
        public void onBindViewHolder(@NonNull MyViewHolder holder, int position) {
            holder.details(position);
        }

        public long getItemId(int position) {
            return position;
        }

        @Override
        public int getItemCount() {
            return originalItems.size();
        }

        @NonNull

        public Filter getFilter() {
            Filter filter = new Filter() {

                @Override
                protected FilterResults performFiltering(CharSequence constraint) {
                    FilterResults results = new FilterResults();
                    final ArrayList<Word> filteredList = new ArrayList<Word>();

                    if (constraint.equals("") || constraint.toString().trim().length() == 0) {
                        results.values = originalItems;
                    } else {
                        String textToFilter = constraint.toString().toLowerCase();
                        for (Word word : originalItems) {
                            if (word.getWord().length() >= textToFilter.length() &&
                                    word.getWord().toLowerCase().contains(textToFilter)) {
                                filteredList.add(word);
                            }
                        }
                        results.values = filteredList;
                    }
                    return results;
                }

                @Override
                protected void publishResults(CharSequence constraint, FilterResults results) {
                    if (results.values != null) {
                        filteredItems = (ArrayList<Word>) results.values;
                        notifyDataSetChanged();
                    }
                }
            };
            return filter;
        }


        class MyViewHolder extends RecyclerView.ViewHolder {
            TextView l1, l2, input, output;
            CardView cardView;
            ImageView copy, share, speak;

            public MyViewHolder(View itemView) {
                super(itemView);
                l1 = itemView.findViewById(R.id.l1);
                l2 = itemView.findViewById(R.id.l2);
                input = itemView.findViewById(R.id.input);
                output = itemView.findViewById(R.id.out);
                cardView = itemView.findViewById(R.id.card);
                copy = itemView.findViewById(R.id.copy);
                share = itemView.findViewById(R.id.share);
                speak = itemView.findViewById(R.id.speak);


            }

            public void details(int position) {
                Word word = originalItems.get(position);
                l1.setText(word.getSourceLanguage());
                l2.setText(word.getTargetLanguage());
                input.setText(word.getWord());
                output.setText(word.getTranslation());
                cardView.setOnLongClickListener(v -> {

                    Toast.makeText(HistoryActivity.this, "Sucessfully Deleted", Toast.LENGTH_SHORT).show();
                    DataBaseHelper dataBaseHelper = new DataBaseHelper(getApplicationContext(), s);
                    dataBaseHelper.deleteWord(word);
                    adapter.notifyItemRemoved(position);
                    adapter.notifyDataSetChanged();
                    recycler.setAdapter(adapter);
                    notifyItemRemoved(position);
                    notifyItemRangeChanged(position, originalItems.size());
                    ArrayList<Word> arrayList = dataBaseHelper.getAllWords();
                    Log.i("123321", "" + arrayList);

                    adapter = new CustomAdapter(getApplicationContext(), arrayList);

                    recycler.setAdapter(adapter);
                    return true;
                });
                copy.setOnClickListener(v -> {
                    ClipboardManager clipboard = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
                    ClipData clip = ClipData.newPlainText("text", word.getTranslation());
                    clipboard.setPrimaryClip(clip);
                    Toast.makeText(HistoryActivity.this, "Successfully copied", Toast.LENGTH_SHORT).show();
                });
                speak.setOnClickListener(v -> {
                    
                    {

                        t1 = new TextToSpeech(getApplicationContext(), new TextToSpeech.OnInitListener() {
                            @Override
                            public void onInit(int status) {
                                if (status != TextToSpeech.ERROR) {
                                    final Locale loc = new Locale(word.getTargetLanguage(), "");

                                    // switch(tts.isLanguageAvailable(loc)){
                                    switch (t1.setLanguage(loc)) {

                                        case TextToSpeech.LANG_AVAILABLE:
                                            Log.i("TAG", "LANG_AVAILABLE");
                                            break;
                                        case TextToSpeech.LANG_COUNTRY_AVAILABLE:
                                            Log.i("TAG", "LANG_COUNTRY_AVAILABLE");
                                            break;
                                        case TextToSpeech.LANG_COUNTRY_VAR_AVAILABLE:
                                            Log.i("TAG", "LANG_COUNTRY_VAR_AVAILABLE");
                                            break;
                                        case TextToSpeech.LANG_MISSING_DATA:
                                            Log.i("TAG", "LANG_MISSING_DATA");
                                            break;
                                        case TextToSpeech.LANG_NOT_SUPPORTED:
                                            Log.i("TAG", "LANG_NOT_SUPPORTED");
                                         //   Toast.makeText(getApplicationContext(), "Language not found", Toast.LENGTH_SHORT).show();
                                            break;
                                    }

                                    t1.speak(
                                            word.getTranslation(), TextToSpeech.QUEUE_FLUSH, null);
                                } else Log.i("123321", "383");
                            }
                        });

                    }
                });
                share.setOnClickListener(v -> {
                    Intent intent = new Intent(Intent.ACTION_SEND);
                    intent.setType("text/plain");
                    intent.putExtra(Intent.EXTRA_TEXT, word.getTranslation());
                    intent.putExtra(Intent.EXTRA_SUBJECT, "Title goes here");
                    startActivity(Intent.createChooser(intent, "Share"));
                });

            }
        }
    }

}

File: app/src/main/java/com/allword/translation/HistoryActivity.kt
package com.allword.translation

import android.app.AlertDialog
import android.content.*
import android.os.Bundle
import android.speech.tts.TextToSpeech
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import com.facebook.ads.*
import kotlinx.android.synthetic.main.activity_history.*
import java.util.*

class HistoryActivity : AppCompatActivity() {
    var adapter: CustomAdapter? = null

    var s: String? = null
    var t1: TextToSpeech? = null

    private var adView: AdView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_history)
        ButterKnife.bind(this)
        val manager = LinearLayoutManager(this)
        manager.reverseLayout = true
        manager.stackFromEnd = true
        recycler!!.layoutManager = manager
        s = intent.getStringExtra("name")
        setSupportActionBar(toolbar)

        adView = AdView(this, "320675462379332_321536232293255", AdSize.BANNER_HEIGHT_50)
        AdSettings.addTestDevice("c77ae079-ddd2-45a2-a632-47ab1cd7f060")
        //

        if (BuildConfig.DEBUG) {
            AdSettings.setTestMode(true)

        }


        val adContainer: LinearLayout = findViewById<View>(R.id.banner_container) as LinearLayout

        adContainer.addView(adView)

        // Request an ad

        // Request an ad
        adView!!.loadAd()
        AdSettings.addTestDevice("c77ae079-ddd2-45a2-a632-47ab1cd7f060");

        tool_title.text = if (s == "Favourites.db") "Bookmark" else "History"
        val dataBaseHelper = DataBaseHelper(applicationContext, s)
        val arrayList = dataBaseHelper.allWords
        Log.i("123321", "" + arrayList)
        adapter = CustomAdapter(applicationContext, arrayList)
        recycler!!.adapter = adapter
        val clear = toolbar!!.findViewById<TextView>(R.id.clear)
        clear.setOnClickListener { v: View? ->
            val builder = AlertDialog.Builder(this)
            builder.setTitle("Are you Sure?")
            builder.setMessage("Delete it permanently")
            builder.setPositiveButton("ok") { dialog: DialogInterface?, which: Int ->
                dataBaseHelper.deleteAllWords()
                val arrayList0 = dataBaseHelper.allWords
                adapter = CustomAdapter(applicationContext, arrayList0)
                recycler!!.adapter = adapter
            }
            builder.setNegativeButton("cancel", null)
            builder.show()
        }
    }

    inner class CustomAdapter(context: Context,
                              objects: ArrayList<Word>) : RecyclerView.Adapter<CustomAdapter.MyViewHolder>() {
        private val originalItems: ArrayList<Word>
        private var filteredItems: ArrayList<Word>
        private val inflater: LayoutInflater
        val count: Int
            get() = filteredItems.size

        fun getItem(position: Int): Word {
            return filteredItems[position]
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
            val view = inflater.inflate(R.layout.list_item, parent, false)
            return MyViewHolder(view)
        }

        override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
            holder.details(position)
        }

        override fun getItemId(position: Int): Long {
            return position.toLong()
        }

        override fun getItemCount(): Int {
            return originalItems.size
        }

        val filter: Filter
            get() = object : Filter() {
                override fun performFiltering(constraint: CharSequence): FilterResults {
                    val results = FilterResults()
                    val filteredList: ArrayList<Word> = ArrayList()
                    if (constraint == "" || constraint.toString().trim { it <= ' ' }.length == 0) {
                        results.values = originalItems
                    } else {
                        val textToFilter = constraint.toString().toLowerCase()
                        for (word in originalItems) {
                            if (word.word.length >= textToFilter.length &&
                                    word.word.toLowerCase().contains(textToFilter)) {
                                filteredList.add(word)
                            }
                        }
                        results.values = filteredList
                    }
                    return results
                }

                override fun publishResults(constraint: CharSequence, results: FilterResults) {
                    if (results.values != null) {
                        filteredItems = results.values as ArrayList<Word>
                        notifyDataSetChanged()
                    }
                }
            }

        inner class MyViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
            var l1: TextView
            var l2: TextView
            var input: TextView
            var output: TextView
            var cardView: CardView
            var copy: ImageView
            var share: ImageView
            var speak: ImageView
            fun details(position: Int) {
                val word = originalItems[position]
                l1.text = word.sourceLanguage
                l2.text = word.targetLanguage
                input.text = word.word
                output.text = word.translation
                cardView.setOnLongClickListener { v: View? ->
                    Toast.makeText(this@HistoryActivity, "Sucessfully Deleted", Toast.LENGTH_SHORT).show()
                    val dataBaseHelper = DataBaseHelper(applicationContext, s)
                    dataBaseHelper.deleteWord(word)
                    adapter!!.notifyItemRemoved(position)
                    adapter!!.notifyDataSetChanged()
                    recycler!!.adapter = adapter
                    notifyItemRemoved(position)
                    notifyItemRangeChanged(position, originalItems.size)
                    val arrayList = dataBaseHelper.allWords
                    Log.i("123321", "" + arrayList)
                    adapter = CustomAdapter(applicationContext, arrayList)
                    recycler!!.adapter = adapter
                    true
                }


                var interstitialAd: InterstitialAd? = null
                AdSettings.addTestDevice("c77ae079-ddd2-45a2-a632-47ab1cd7f060")
                //

                if (BuildConfig.DEBUG) {
                    AdSettings.setTestMode(true)

                }
                interstitialAd = InterstitialAd(copy.context, "320675462379332_321528522294026")

                interstitialAd!!.loadAd()





                copy.setOnClickListener {


                    interstitialAd!!.setAdListener(object : InterstitialAdListener {
                        override fun onInterstitialDisplayed(ad: Ad) {
                            // Interstitial ad displayed callback
                            Log.e("123321", "Interstitial ad displayed.")
                        }

                        override fun onInterstitialDismissed(ad: Ad) {
                            // Interstitial dismissed callback
                            Log.e("123321", "Interstitial ad dismissed.")




interstitialAd.loadAd()
                            val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
                            val clip = ClipData.newPlainText("text", word.translation)

                            clipboard.setPrimaryClip(clip)
                            Toast.makeText(this@HistoryActivity, "Successfully copied", Toast.LENGTH_SHORT).show()


                        }

                        override fun onError(ad: Ad?, adError: AdError) {
                            // Ad error callback
                            Log.e("123321", "Interstitial ad failed to load: " + adError.getErrorMessage())
                        }

                        override fun onAdLoaded(ad: Ad) {
                            // Interstitial ad is loaded and ready to be displayed
                            Log.d("123321", "Interstitial ad is loaded and ready to be displayed!")
                            // Show the ad

                        }

                        override fun onAdClicked(ad: Ad) {
                            // Ad clicked callback
                            Log.d("123321", "Interstitial ad clicked!")
                        }

                        override fun onLoggingImpression(ad: Ad) {
                            // Ad impression logged callback
                            Log.d("123321", "Interstitial ad impression logged!")
                        }
                    })

                    if(interstitialAd!!.isAdLoaded)
                    {interstitialAd!!.show()
                        interstitialAd!!.loadAd()

                    }

                    else

                 {   val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
                    val clip = ClipData.newPlainText("text", word.translation)

                    clipboard.setPrimaryClip(clip)
                    Toast.makeText(this@HistoryActivity, "Successfully copied", Toast.LENGTH_SHORT).show()}
                }
                speak.setOnClickListener { v: View? ->
                    run {
                        t1 = TextToSpeech(applicationContext, TextToSpeech.OnInitListener { status ->
                            if (status != TextToSpeech.ERROR) {
                                val loc = Locale(word.targetLanguage, "")
                                when (t1!!.setLanguage(loc)) {
                                    TextToSpeech.LANG_AVAILABLE -> Log.i("TAG", "LANG_AVAILABLE")
                                    TextToSpeech.LANG_COUNTRY_AVAILABLE -> Log.i("TAG", "LANG_COUNTRY_AVAILABLE")
                                    TextToSpeech.LANG_COUNTRY_VAR_AVAILABLE -> Log.i("TAG", "LANG_COUNTRY_VAR_AVAILABLE")
                                    TextToSpeech.LANG_MISSING_DATA -> Log.i("TAG", "LANG_MISSING_DATA")
                                    TextToSpeech.LANG_NOT_SUPPORTED -> Log.i("TAG", "LANG_NOT_SUPPORTED")
                                }
                                t1!!.speak(
                                        word.translation, TextToSpeech.QUEUE_FLUSH, null)
                            } else Log.i("123321", "383")
                        })
                    }
                }
                share.setOnClickListener { v: View? ->
                    val intent = Intent(Intent.ACTION_SEND)
                    intent.type = "text/plain"
                    intent.putExtra(Intent.EXTRA_TEXT, word.translation)
                    intent.putExtra(Intent.EXTRA_SUBJECT, "Title goes here")
                    startActivity(Intent.createChooser(intent, "Share"))
                }
            }

            init {
                l1 = itemView.findViewById(R.id.l1)
                l2 = itemView.findViewById(R.id.l2)
                input = itemView.findViewById(R.id.input)
                output = itemView.findViewById(R.id.out)
                cardView = itemView.findViewById(R.id.card)
                copy = itemView.findViewById(R.id.copy)
                share = itemView.findViewById(R.id.share)
                speak = itemView.findViewById(R.id.speak)
            }
        }

        init {
            originalItems = ArrayList()
            originalItems.addAll(objects)
            filteredItems = ArrayList()
            filteredItems.addAll(originalItems)
            inflater = LayoutInflater.from(context)
        }
    }
}